### PR TITLE
[fix][minor] ShardedDDP train/eval modes

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
ShardedDDP was not respecting the eval() call from the base nn.Module class, at least not if the asserts were allowed

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.
cc @vedanuj 
## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 1]
```diff
@@ -352,7 +352,7 @@ def _clear_counters(self) -> None:
             assert self._bucket_list is not None
 
             for bucket in self._bucket_list:
-                assert self.should_accumulate_grads or bucket.sent, (
+                assert not self.training or self.should_accumulate_grads or bucket.sent, (
                     "A bucket failed to be sent, probably unused parameters."
```
Comment (anj-s) [inline fairscale/nn/data_parallel/sharded_ddp.py:356]: Should we update this message?

File: tests/nn/data_parallel/test_sharded_ddp.py [hunk 2]
```diff
@@ -440,6 +440,34 @@ def test_device_change():
...
 
+def run_test_training_change(rank, world_size, backend, device, temp_file_name):
+    url = "file://" + temp_file_name
+    group = dist.init_process_group(init_method=url, backend=backend, rank=rank, world_size=world_size)
+    model = Sequential(Linear(2, 3), Linear(3, 3)).to(device)
+    optimizer = OSS(params=model.parameters(), optim=torch.optim.SGD, lr=0.01, momentum=0.99)
+    ddp_model = ShardedDataParallel(model, optimizer, process_group=group)
+
+    inputs = torch.rand((10, 2), device=device)
+    outputs = ddp_model(inputs)  # assert if the module has not been changed properly
```
Comment (anj-s) [inline tests/nn/data_parallel/test_sharded_ddp.py:451]: just to clarify, this would have thrown an error before right?

File: tests/nn/data_parallel/test_sharded_ddp.py [hunk 3]
```diff
@@ -440,6 +440,34 @@ def test_device_change():
...
+    _ = outputs.norm().backward()
+
+    ddp_model.eval()
+    ddp_model(inputs)  # This will assert if eval() is not properly taken into account
```
Comment (anj-s) [inline tests/nn/data_parallel/test_sharded_ddp.py:455]: just to clarify, this would have thrown an error before right?

File: tests/nn/data_parallel/test_sharded_ddp.py [hunk 4]
```diff
@@ -440,6 +440,34 @@ def test_device_change():
...
+    ddp_model(inputs)
+
+    dist.destroy_process_group()
+
+
+@skip_if_no_cuda
```
Comment (anj-s) [inline tests/nn/data_parallel/test_sharded_ddp.py:461]: Are we using GPUs in this test? Given its gloo can we avoid skipping?

**Review by anj-s — APPROVED** [review]

File: fairscale/nn/data_parallel/sharded_ddp.py (hunk 1 above)
Comment (blefaudeux) [reply fairscale/nn/data_parallel/sharded_ddp.py:356]: I think it's ok right now, basically this assert would throw before when in evaluation mode, because the .eval() call was not properly caught
